### ISSUE TITLE
Value class handling

### ISSUE THREAD

**Issue by mpollmeier** [netvl/picopickle#16]
By default a value class gets serialised to a Map with one entry, which is a reasonable default, as it behaves just like any other case class:

```
import io.github.netvl.picopickle.backends.collections.CollectionsPickler._
case class MyValueClass(value: String) extends AnyVal
write(MyValueClass("someValue"))
// Map(value -> someValue)
```

I would like to serialise it to/from a single value only though, i.e. the above `write` should just return "someValue" and `read` should construct it from the value (basically just ignore the key). 

One option might be to have a read/write implementation for case classes that happen to be AnyVal. It needs to have precedence over normal case classes. 

```
case class MyValueClass(value: String) extends AnyVal
case class NormalCC(i: Int)

def valueClassWriter[A <: AnyVal: shapeless.Generic](a: A): Unit = ???

valueClassWriter(MyValueClass("someValue")) //compiles
valueClassWriter(NormalCC(55)) // does not compile
valueClassWriter(34) // does not compile
```

The implementation would probably need reflection and I've done some groundwork here: https://github.com/mpollmeier/shapeless-playground/blob/9254bf5bac28d4f55800c13419a09478d3268599/MapReaderWriter.scala#L80:L100

Does that make sense? 
What's the best place to add this functionality?

**Comment by mpollmeier** [comment]
updated the above to ensure the method can't be called with primitives, using shapeless.Generic

**Comment by netvl** [comment]
Yep, it does make sense. And it probably doesn't even need reflection. I think it should be possible to add a new pair of traits (like `ShapelessReaders` and `ShapelessWriters`, merged into a single `ShapelessReaderWriters`) which would contain something like

``` scala
implicit def valueClassReader[T <: AnyVal, R](implicit gen: Generic.Aux[T, R :: HNil],
                                              rr: Reader[R]): Reader[T] =
  rr.andThen(r => gen.from(r :: HNil))
```

and

``` scala
implicit def valueClassWriter[T <: AnyVal, R](implicit gen: Generic.Aux[T, R :: HNil],
                                              rw: Writer[R]): Writer[T] =
  Writer(t => gen.to(t) match {
    case r :: HNil => rw.write(r)
  })
```

They shouldn't be mixed by default to keep backwards compatibility. But if one wants to opt into this behavior, they could mix this trait into their pickler. I think implicit priorities should be just right in such scenario. I would accept a pull request for this, so if you want you can do it :) I'll probably add it myself if no one is willing to do it, but I can't guarantee any timelines, unfortunately.

**Comment by mpollmeier** [comment]
Thanks for those pointers, and indeed there's no need to use reflection here - awesome! I got it working here: 
https://github.com/mpollmeier/picopickle-playground/blob/62c5df687bd9f184cf6a9a16dfdc7a8542a761ce/src/test/scala/PicklerTest.scala#L56:L90

Two more questions: 
- for some reason the `implicit gen: Generic.Aux[T, R :: HNil]` is not being found, I had to use two type classes instead of one (`gen` and `hasOne`). Any idea what I'm doing wrong?
- I had to cast to the VCAsHList... I know it's the right type, how can I convince the compiler?

Afterwards I'll wrap it up and send you a PR. 
Btw I'm planning on using this in https://github.com/mpollmeier/gremlin-scala

**Comment by netvl** [comment]
> for some reason the implicit gen: Generic.Aux[T, R :: HNil] is not being found, I had to use two type classes instead of one (gen and hasOne). Any idea what I'm doing wrong?

Indeed, it does not work, probably because of type resolution. I guess that scalac, when typechecking the use site of the reader, does not unify the representation type with whatever the macro-generated `Generic.Repr` is. It seems that `IsHCons` is the correct solution.

> I know it's the right type, how can I convince the compiler?

I've experimented with it and it seems that it is possible to request `Repr =:= (R :: HNil)` evidence which can then be used for safe type cast. For example, the following works:

``` scala
import shapeless._
import shapeless.ops.hlist.IsHCons

def extract[T, Repr <: HList, R](x: T)(implicit gen: Generic.Aux[T, Repr],
                                       hc: IsHCons.Aux[Repr, R, HNil],
                                       eqv: Repr =:= (R :: HNil)): (R :: HNil, R) =
  (eqv(gen.to(x)), hc.head(gen.to(x)))
```

You may need to swap the arguments of `=:=`, however.

> Afterwards I'll wrap it up and send you a PR. 

Thanks!

> Btw I'm planning on using this

Great to hear, finally it gets used somewhere :)

**Comment by mpollmeier** [comment]
Spot on, that worked. Shapeless is simply awesome. 
And yes, I had to swap the arguments. 
PR coming shortly. Thank you!

**Comment by netvl** [comment]
Value classes serialization is now available in version 0.3.0.
